### ISSUE TITLE
One-hot operation causes super high workItem count and causes errors

### ISSUE THREAD

**Issue by jarek108** [Unity-Technologies/barracuda-release#243]
When we perform a 10 class one-hot operation on a flattened 512x512 grayscale image (zdim = 262144) a warning ocures: "Thread group count is above the maximum allowed limit. Maximum allowed thread group count is 65535." 

it is followed by an error: "Exceeded safe compute dispatch group count limit per dimension [1, 1, 65536] for Transpose_NHWC"

it seems one workItem is required per each of the 262144 values, which kills the gpu. this perhaps seems excessive or just wrong?

the model works fine in pytorch, we are unsure if we are doing sth wrong or is this a bug?

here is the mentioned model: https://drive.google.com/file/d/1osYj7_C9ZSssTXnylZOLN4rkgqFt5Xvj/view?usp=sharing

**Comment by AlexRibard** [comment]
Hello @jarek108 the 65535 thread group limit is given by the HLSL API, so with current implementations, it means that your input is too large to fit within dispatch budget.
The solution is for us to unroll the dispatch over a few dimensions. We already do that for activations and a few other kernel.
I'll take a look at your model and see if there is anything wrong with the import side of it.
In the mean time, is reducing the input size a possibility?

**Comment by jarek108** [comment]
@AlexRibard thank you for fast reaction! it is not possible for us to reduce the general size from 512x512x1, however, perhaps it would be possible to achieve the one-hot-like behavior on a (1, 10, 512,512) tensor instead of the (1, 10, 1, 512*512) tensor, as we are attempting to do now (basically we want 10 class-related tensors that have 1s only in the pixels where a specific class is present and we do not care if it is flattened or unflattened 512x512 tensor. we also want to produce this from an input 512x512 grayscale image where pixel value encodes the class id). however, we do not know how to do this in a way that meets current restrictions of barracuda onnx. perhaps you can advise how we should alter our model after looking at   it?

**Comment by AlexRibard** [comment]
yes, if you can split the input yourself into 1,10,512,512 that would work. It would be what we would be doing under the hood for the dispatch.
`OneHot` has a `axis` parameter. So I think if the input is (1, 10, 512,512) and you OneHot on axis=1 I think it should do as you expect

**Comment by kdziedzic66** [comment]
@AlexRibard Thanks for your replays!

We are trying do dispatch one hot operation now on two-dimensional tensor **input** (1, 1, 512, 512) - the output **seg** should be (1,182,512,512). we are using the following python code to achive that:

```
seg = torch.nn.functional.one_hot(input, 182).transpose(1, 4).squeeze(-1)
seg = seg.reshape(1,182,512,512)
seg = seg.float()
```
those operations are part of `forward` method in our pytorch module. 
The export to onnx works fine: pytorch version is 1.10, onnx version 1.10.2.

After importing onnx model to Barracuda we are geting the following error:

`Unexpected error while evaluating model output 497. System.ArgumentException: Cannot reshape array of size 93184 into shape (n:1, h:182, w:512, c:512)`

It looks like one hot operation gives slightly different output format than the one in pytorch.  Do you have any ideas how to solve this? The onnx exported model is here: https://drive.google.com/file/d/1MQBRo947ViZDj87kqB_aPcqeSIkabiJ8/view?usp=sharing

**Comment by AlexRibard** [comment]
Ah, yes it's probably our shape inference messing up. Let me take a look at the model then.
Thanks for providing it

**Comment by AlexRibard** [comment]
Ok I took a look at your model.
I got it to run by modifying the input dimension to 1x512x512 (4D -> 5D OneHot is not supported atm).
(I also needed to modify the transpose to be a 4D transpose).
There is some typo so you'll get an error when importing it, but it runs ok. I will push the fix, but I can share it to you if you need it.

There is another potential issue. You use float `Resize` which we don't support. So it will be rounded down to 0 and thus not do anything.
Could you use `AveragePool` instead?

**Comment by kdziedzic66** [comment]
@AlexRibard  Thanks for your help!

We have another two questions: 

1. Where did you push your fix? Could you share it with us?
2. We are trying to solve the same thing using `Scatter` instead of `OneHot` and it throws an error - it looks like `ScatterND` (generalization of `Scatter`) is supported and `Scatter` is not. Would it be possible in the future to support `Scatter` too? We have attempted to solve this by changing the operator type to `ScatterND` when the type was `Scatter` however it still doesn't work. We've been trying with two models: 
            a.  https://drive.google.com/file/d/1XDgj3vMkjk3dKiUhohbYNz9TCgQ7zDxs/view?usp=sharing (it uses `torch.Tensor.scatter_` method: https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html) - here is the full error stack trace: ```ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) (at <695d1cc93cca45069c528c15c9fdd749>:0)
System.ThrowHelper.ThrowArgumentOutOfRangeException () (at <695d1cc93cca45069c528c15c9fdd749>:0)
System.Collections.Generic.List`1[T].get_Item (System.Int32 index) (at <695d1cc93cca45069c528c15c9fdd749>:0)
Unity.Barracuda.Compiler.IRShapeInferenceHelper.ShapeInference.InferOutputShapeNCHW (Unity.Barracuda.Layer layer, System.Nullable`1[System.Int32][] inputRanks, System.Nullable`1[Unity.Barracuda.TensorShape][] inputShapes) (at C:/Users/Jarek/PycharmProjects/pythonProject/barracuda-release/Barracuda/Runtime/Core/Compiler/ShapeInference/IRShapeInferenceHelper.cs:667)
Unity.Barracuda.Compiler.IRShapeInferenceHelper.ShapeInference.UpdateKnownTensorShapesNCHW (Unity.Barracuda.Model model, System.Collections.Generic.IDictionary`2[System.String,System.Nullable`1[System.Int32]]& ranksByName, System.Collections.Generic.IDictionary`2[System.String,System.Nullable`1[Unity.Barracuda.TensorShape]]& shapesByName) (at C:/Users/Jarek/PycharmProjects/pythonProject/barracuda-release/Barracuda/Runtime/Core/Compiler/ShapeInference/IRShapeInferenceHelper.cs:908)
Unity.Barracuda.Compiler.Passes.IRShapeInferenceAndConstantFusing.FuseShapesIntoConstants (Unity.Barracuda.Model& model, System.Collections.Generic.IDictionary`2[TKey,TValue] shapesByName, System.Collections.Generic.IDictionary`2[TKey,TValue] ranksByName, System.Collections.Generic.List`1[Unity.Barracuda.Model+ImporterWarning]& warnings) (at C:/Users/Jarek/PycharmProjects/pythonProject/barracuda-release/Barracuda/Runtime/Core/Compiler/Passes/IRShapeInferenceAndConstantFusing.cs:73)
Unity.Barracuda.Compiler.Passes.IRShapeInferenceAndConstantFusing.Run (Unity.Barracuda.Model& model, System.Collections.Generic.List`1[T] warnings) (at C:/Users/Jarek/PycharmProjects/pythonProject/barracuda-release/Barracuda/Runtime/Core/Compiler/Passes/IRShapeInferenceAndConstantFusing.cs:30)
Unity.Barracuda.Compiler.Passes.IntermediateToRunnableNHWCPass.Run (Unity.Barracuda.Model& model) (at C:/Users/Jarek/PycharmProjects/pythonProject/barracuda-release/Barracuda/Runtime/Core/Compiler/Passes/IntermediateToRunnableNHWCPass.cs:15)
Unity.Barracuda.ONNX.ONNXModelConverter.Convert (Google.Protobuf.CodedInputStream inputStream) (at C:/Users/Jarek/PycharmProjects/pythonProject/barracuda-release/Barracuda/Runtime/ONNX/ONNXModelConverter.cs:188)
Unity.Barracuda.ONNX.ONNXModelConverter.Convert (System.Byte[] buffer) (at C:/Users/Jarek/PycharmProjects/pythonProject/barracuda-release/Barracuda/Runtime/ONNX/ONNXModelConverter.cs:109)
Toolkit.LoadOnnxModel (System.Byte[] rawModel) (at Assets/Scripts/Toolkit.cs:40)
BarracudaInference.LoadModelFromFile (System.String relativePath, System.Boolean forceLoad) (at Assets/Scripts/BarracudaInference.cs:301)
BarracudaInference.InitGanModel (System.String modelPath, System.Boolean forceLoad) (at Assets/Scripts/BarracudaInference.cs:155)
BarracudaInference_Editor.OnInspectorGUI () (at Assets/Editor/BarracudaInference_Editor.cs:19)
UnityEditor.UIElements.InspectorElement+<>c__DisplayClass59_0.<CreateIMGUIInspectorFromEditor>b__0 () (at <2758fd44916644e3a3b65c9d4809b7f9>:0)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr, Boolean&)```
             b.  https://drive.google.com/file/d/1IQlvnC6KBIH43CoW_QjFLr31YUyEPFeg/view?usp=sharing (uses `torch.Tensor.scatter_add`https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html)  - in this case we are getting an error:        
![unity_error](https://user-images.githubusercontent.com/12718383/148199176-b7388153-4373-4b5d-ad4b-63971fbac07d.png)

**Comment by AlexRibard** [comment]
Here is the modified top part of the model
![image](https://user-images.githubusercontent.com/12990800/148238252-865e91e8-943f-461e-b8e2-e8ba7cc463dd.png)
And the corresponding code I used to edit it

```
import numpy as np
import onnx
from onnx import ModelProto, TensorProto, numpy_helper, helper, external_data_helper, save_model

model = onnx.load("gaugan_landscapes_grayscale_input_one_hot.onnx")

graph = model.graph

for node in graph.node:
    if(node.name == "Transpose_3"):
        #print(node.attribute[0])
        node.attribute[0].ints[0] = 0
        node.attribute[0].ints[1] = 3
        node.attribute[0].ints[2] = 1
        node.attribute[0].ints[3] = 2
        node.attribute[0].ints.pop(4)
    if(node.name == "Squeeze_4"):
        node.op_type = 'Identity'
        node.attribute.pop()
    if(node.name == "Reshape_6"):
        node.op_type = 'Identity'


model.graph.input[0].type.tensor_type.shape.dim.pop(0)
    
onnx.save(model, "gaugan_landscapes_grayscale_input_one_hot_2.onnx")
```

OneHot should work. Inference runs without error on latest. The fix is the following:
![image](https://user-images.githubusercontent.com/12990800/148238755-8dc9b015-a4db-4591-9d98-5eb7508cf123.png)
Backends/ModelAnalyser.cs

But again there is the issues with the Resize that I mentioned before

**Comment by FlorentGuinier** [comment]
Closing this issue as there is/has been no activity for some time. Please reopen if needed.
